// @redwoodjs/api/dist/cache/clients/MemcachedClient.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _promise = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/promise"));
var _stringify = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/json/stringify"));
var _interopRequireWildcard2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/interopRequireWildcard"));
var _BaseClient = _interopRequireDefault(require("./BaseClient"));
class MemcachedClient extends _BaseClient.default {
  constructor(servers, options) {
    super();
    this.client = void 0;
    this.servers = void 0;
    this.options = void 0;
    this.servers = servers;
    this.options = options;
  }
  async connect() {
    const {
      Client: MemCachedClient
    } = await _promise.default.resolve().then(() => (0, _interopRequireWildcard2.default)(require('memjs')));
    this.client = MemCachedClient.create(this.servers, this.options);
  }
  async disconnect() {
    var _this$client;
    (_this$client = this.client) === null || _this$client === void 0 ? void 0 : _this$client.close();
    this.client = null;
  }
  async get(key) {
    var _this$client2;
    if (!this.client) {
      await this.connect();
    }
    const result = await ((_this$client2 = this.client) === null || _this$client2 === void 0 ? void 0 : _this$client2.get(key));
    if (result !== null && result !== void 0 && result.value) {
      return JSON.parse(result.value.toString());
    } else {
      return result === null || result === void 0 ? void 0 : result.value;
    }
  }
  async set(key, value, options) {
    var _this$client3;
    if (!this.client) {
      await this.connect();
    }
    return (_this$client3 = this.client) === null || _this$client3 === void 0 ? void 0 : _this$client3.set(key, (0, _stringify.default)(value), options);
  }
}
exports.default = MemcachedClient;
